### PR TITLE
[ROCm] Add native MXFP4 (W4A4) MoE backend for minimax-m3

### PR DESCRIPTION
**Pull request by YukioZzz**
MiniMax-M3 uses W4A4 MXFP4 experts with the uninterleaved OAI clamped-
SwiGLU (SWIGLUOAI_UNINTERLEAVE). Wire it onto AITER's existing W4A4
per_1x32 ck_tile MoE kernel on gfx950 (no new kernel):

- rocm_aiter_moe: map SWIGLUOAI_UNINTERLEAVE to AITER "swiglu" and add it
  to the supported activations.
- oracle/mxfp4: select the W4A4 weight-prep layout by activation. SwiGLU-OAI
  dispatches to the ck_tile FlatMM kernel and needs the gate/up interleave-16
  weight+scale layout (shuffle_weight_a16w4 / shuffle_scale_a16w4); other
  activations keep the classic CK 2-stage layout (generic shuffle_weights +
  e8m0). The old code applied the classic layout unconditionally, which
  mispaired gate/up for M3 (SwiGLU -> ck_tile) and produced garbage. M3
  stores gate/up separated, so no row de-interleave.

Additive to the AITER MXFP4 path; W4A16 (gpt-oss) and non-SwiGLU W4A4
models keep their existing layout. Validated on MiniMax-M3 (gfx950):
gsm8k 5-shot exact_match 0.95, matching the BF16 emulation baseline.

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/oracle/mxfp4.py [hunk 1]
```diff
@@ -503,7 +515,22 @@ def select_mxfp4_moe_backend(
...
+        # SwiGLU-OAI is served natively by TRITON_NATIVE_MX in the loop above.
+        from vllm.model_executor.layers.fused_moe.experts.ocp_mx_emulation_moe import (
+            OCP_MXQuantizationEmulationTritonExperts,
+        )
+
+        logger.info_once(
+            "No native MXFP4 MoE backend matched on this ROCm device; falling "
+            "back to BF16 emulation (weights dequantized once at load)."
+        )
+        return Mxfp4MoeBackend.EMULATION, OCP_MXQuantizationEmulationTritonExperts
```
Comment (BowenBao) [inline vllm/model_executor/layers/fused_moe/oracle/mxfp4.py]: Please revert to original behavior. Let's avoid accidentally falling back to slow performance emulation.

File: vllm/model_executor/layers/fused_moe/experts/mxfp4_native_moe.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+from vllm.triton_utils import tl, triton
+
+logger = init_logger(__name__)
+
+# e2m1 grid midpoints for nearest-rounding (positive magnitudes
+# 0,0.5,1,1.5,2,3,4,6 -> 7 split points). Lazily moved to device.
+_E2M1_MIDS = torch.tensor([0.25, 0.75, 1.25, 1.75, 2.5, 3.5, 5.0])
+
+
+def mxfp4_quantize(x: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
```
Comment (BowenBao) [inline vllm/model_executor/layers/fused_moe/experts/mxfp4_native_moe.py]: @fxmarty-amd please help review

File: vllm/model_executor/layers/quantization/quark/quark_moe.py [hunk 3]
```diff
@@ -1093,9 +1093,9 @@ def maybe_roundup_sizes(
         )
         # In case quantization emulation backend is used, there is no need to apply
         # MXFP4-specific padding logic as the compute happens in higher precision.
-        if (
-            self.mxfp4_backend is not None
-            and self.mxfp4_backend != Mxfp4MoeBackend.EMULATION
+        if self.mxfp4_backend is not None and self.mxfp4_backend not in (
+            Mxfp4MoeBackend.EMULATION,
+            Mxfp4MoeBackend.TRITON_NATIVE_MX,
```
Comment (BowenBao) [inline vllm/model_executor/layers/quantization/quark/quark_moe.py]: I'd recommend avoid this change and keep same padding logic. 
1. it supports more model and tp size combination where size is not divisible without padding.
2. padding should provide better perf based on experience.

**Review by BowenBao — COMMENTED** [review]
why not update aiter ck kernel to support new activation function?

File: vllm/model_executor/layers/quantization/quark/quark_moe.py (hunk 3 above)
Comment (YukioZzz) [reply vllm/model_executor/layers/quantization/quark/quark_moe.py]: Done, quark_moe untouched.

**Comment by YukioZzz** [comment]
> why not update aiter ck kernel to support new activation function?

Done, now the existed native aiter ck_tile kernel is used.

File: vllm/model_executor/layers/fused_moe/oracle/mxfp4.py (hunk 1 above)
Comment (YukioZzz) [reply vllm/model_executor/layers/fused_moe/oracle/mxfp4.py]: Done.

File: vllm/model_executor/layers/fused_moe/oracle/mxfp4.py [hunk 4]
```diff
@@ -962,12 +962,55 @@ def swap_every_two_rows(x, axis=-1):
...
 
+        # AITER fused_moe dispatches W4A4 + SwiGLU-OAI (per_1x32) to the ck_tile
+        # FlatMM kernel, which needs the gate/up interleave-16 weight+scale
+        # layout (shuffle_*_a16w4; activation-dtype independent). Other
+        # activations use the classic CK 2-stage kernel (generic shuffle_weights
+        # + e8m0 scale swizzle). Select the layout by activation so both paths
+        # work.
+        if layer.activation in (
+            MoEActivation.SWIGLUOAI,
+            MoEActivation.SWIGLUOAI_UNINTERLEAVE,
```
Comment (BowenBao) [inline vllm/model_executor/layers/fused_moe/oracle/mxfp4.py:980]: why is interleave needed below for the activation that has `UNINTERLEAVE` in the name `SWIGLUOAI_UNINTERLEAVE`?

edit: also why can `MoEActivation.SWIGLUOAI` share same code path with `MoEActivation.SWIGLUOAI_UNINTERLEAVE`? from name they should differ in interleave setting

**Review by BowenBao — COMMENTED** [review]
also seems duplicate of #46419?

File: vllm/model_executor/layers/fused_moe/oracle/mxfp4.py (hunk 4 above)
Comment (YukioZzz) [reply vllm/model_executor/layers/fused_moe/oracle/mxfp4.py:980]: You are right, here we should only include SWIGLUOAI_UNINTERLEAVE.

**Comment by YukioZzz** [comment]
> also seems duplicate of #46419?

Ah, alright. Seems AITER will deliver a new un-interleaved kernel, then no need to preprocess in the vllm side.

**Comment by YukioZzz** [comment]
duplicate of https://github.com/vllm-project/vllm/pull/46419
